README.md
<p align="center">
  <img src="https://i.imgur.com/lNdBu1v.png" alt="onedark.nvim" />
  <br />
  Atom's iconic One Dark theme for Neovim, written in Lua
</p>

<p align="center">
  <a href="https://github.com/monsonjeremy/onedark.nvim">
    <img alt="monsonjeremy/onedark.nvim" src="https://img.shields.io/badge/Fork%20of-monsonjeremy%2Fonedark.nvim-blue" />
  </a>
  <a href="https://github.com/ful1e5/onedark.nvim/actions">
    <img alt="GitHub Action Linting" src="https://github.com/ful1e5/onedark.nvim/actions/workflows/lint.yml/badge.svg" />
  </a>
  <a href="twitter.com/ful1e5">
    <img alt="Twitter" src="https://img.shields.io/badge/twitter-ful1e5-blue" />
  </a>
</p>

## Features

- Supports the latest Neovim 0.5 features like TreeSitter and LSP
- Minimal inactive StatusLine
- Vim terminal colors
- Darker background for sidebar-like windows
- Color configs for [Kitty](https://sw.kovidgoyal.net/kitty/conf.html?highlight=include) and [Alacritty](https://github.com/alacritty/alacritty)
- Beautiful **lualine** theme

## Requirements

- Neovim >= 0.5.0

## Installation

Install the theme with your preferred package manager:

[vim-plug](https://github.com/junegunn/vim-plug)

```vim
Plug 'ful1e5/onedark.nvim'
```

[packer](https://github.com/wbthomason/packer.nvim)

```lua
use 'ful1e5/onedark.nvim'
```

## Usage

Enable the colorscheme:

```vim
" Vim Script
colorscheme onedark
```

```lua
-- Lua
require('onedark').setup()
```

## Configuration

| Option                   | Default    | Description                                                                                                                                                                       |
| ------------------------ | ---------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| colors                   | `{}`       | You can override specific color groups to use other groups or a hex color.                                                                                                        |
| comment_style            | `italic`   | Highlight style for comments (check `:help highlight-args` for options)                                                                                                           |
| dark_float               | `false`    | Float windows like the lsp diagnostics windows get a darker background.                                                                                                           |
| dark_sidebar             | `true`     | Sidebar like windows like `NvimTree` get a darker background.                                                                                                                     |
| dev                      | `false`    | Developer Mode.                                                                                                                                                                   |
| function_style           | `NONE`     | Highlight style for functions (check `:help highlight-args` for options)                                                                                                          |
| hide_end_of_buffer       | `true`     | Enabling this option, will hide filler lines (~) after the end of the buffer.                                                                                                     |
| hide_inactive_statusline | `true`    | Enabling this option, will hide inactive statuslines and replace them with a thin border instead. Should work with the standard **StatusLine**.                                   |
| highlight_linenumber     | `false`    | Enabling this option, will enable dark color to `LineNr`, `SignColumn` and `CursorLineNr` highlights.(also support [gitsigns](https://github.com/lewis6991/gitsigns.nvim) plugin) |
| keyword_style            | `italic`   | Highlight style for keywords (check `:help highlight-args` for options)                                                                                                           |
| lualine_bold             | `false`    | When `true`, section headers in the lualine theme will be bold.                                                                                                                   |
| msg_area_style           | `NONE`     | Highlight style for messages and cmdline (check `:help highlight-args` for options)                                                                                               |
| overrides                | `function` | Override specific highlight groups. The function accpet colors as argument. You can also add a non-exists highlight by enabling the `dev` mode.                                   |
| sidebars                 | `{}`       | Set a darker background on sidebar-like windows. For example: `{"qf", "vista_kind", "terminal", "packer"}`                                                                        |
| transparent              | `false`    | Enable this to disable setting the background color.                                                                                                                              |
| transparent_sidebar      | `false`    | Sidebar like windows like `NvimTree` get a transparent background.                                                                                                                |
| variable_style           | `NONE`     | Highlight style for variables and identifiers (check `:help highlight-args` for options)                                                                                          |

```vim
" Example config in VimScript
" configuration needs to be set BEFORE loading the color scheme with `colorscheme` command
let g:onedark_function_style = "italic"
let g:onedark_sidebars = ["qf", "vista_kind", "terminal", "packer"]

" Change the "hint" color to the "orange0" color, and make the "error" color bright red
let g:onedark_colors = {
  \ 'hint': 'orange0',
  \ 'error': '#ff0000'
\ }

" Load the colorscheme
colorscheme onedark
```

```lua
-- Example config in Lua
require("onedark").setup({
  function_style = "italic",
  sidebars = {"qf", "vista_kind", "terminal", "packer"},

  -- Change the "hint" color to the "orange0" color, and make the "error" color bright red
  colors = {hint = "orange0", error = "#ff0000"},

  -- Overwrite the highlight groups
  overrides = function(c)
    return {
      htmlTag = {fg = c.red0, bg = "#282c34", sp = c.hint, style = "underline"},
      DiagnosticHint = {link = "LspDiagnosticsDefaultHint"},
      -- this will remove the highlight groups
      TSField = {},
    }
  end
})
```

### Lualine Support

To enable the `onedark` theme for `Lualine`, simply specify it in your lualine settings:

```lua
require('lualine').setup {
  options = {
    theme = 'onedark-nvim',
    -- ... your lualine config
  }
}
```

![normal](https://imgur.com/utoeRWg.png)
![insert](https://imgur.com/cimWSSb.png)
![visual](https://imgur.com/4ntUbvB.png)
![command](https://imgur.com/bmreFgJ.png)
![terminal](https://imgur.com/Fw0Btae.png)

## Terminal Themes

> To generate the configs `make terminal` or `:luafile lua/onedark/terminal/init.lua`

Extra color configs for **kitty**, and **Alacritty** can be found in [terminal](terminal/) directory. To use them, refer to their respective documentation.

- kitty - [./terminal/kitty](https://github.com/ful1e5/onedark.nvim/tree/main/terminal/kitty)
- Alacritty - [./terminal/alacritty](https://github.com/ful1e5/onedark.nvim/tree/main/terminal/alacritty)

### Making `undercurls` work properly in **Tmux**

To have undercurls show up and in color, add the following to your **Tmux** config file:

```sh
# Undercurl
set -g default-terminal "${TERM}"
set -as terminal-overrides ',*:Smulx=\E[4::%p1%dm'  # undercurl support
set -as terminal-overrides ',*:Setulc=\E[58::2::%p1%{65536}%/%d::%p1%{256}%/%{255}%&%d::%p1%{255}%&%d%;m'  # underscore colours - needs tmux-3.0
```

## Plugin Support

- [Builtin LSP diagnostics](https://neovim.io/doc/user/lsp.html)
- [TimUntersberger/neogit](https://github.com/TimUntersberger/neogit)
- [airblade/vim-gitgutter](https://github.com/airblade/vim-gitgutter)
- [akinsho/nvim-bufferline.lua](https://github.com/akinsho/nvim-bufferline.lua)
- [folke/lsp-trouble.nvim](https://github.com/folke/lsp-trouble.nvim)
- [ggandor/lightspeed.nvim](https://github.com/ggandor/lightspeed.nvim)
- [glepnir/dashboard-nvim](https://github.com/glepnir/dashboard-nvim)
- [glepnir/lspsaga.nvim](https://github.com/glepnir/lspsaga.nvim)
- [hrsh7th/nvim-cmp](https://github.com/hrsh7th/nvim-cmp)
- [kyazdani42/nvim-tree.lua](https://github.com/kyazdani42/nvim-tree.lua)
- [lambdalisue/glyph-palette.vim](https://github.com/lambdalisue/glyph-palette.vim)
- [lewis6991/gitsigns.nvim](https://github.com/lewis6991/gitsigns.nvim)
- [liuchengxu/vim-which-key](https://github.com/liuchengxu/vim-which-key)
- [lukas-reineke/indent-blankline.nvim](https://github.com/lukas-reineke/indent-blankline.nvim)
- [neoclide/coc.nvim](https://github.com/neoclide/coc.nvim)
- [nvim-lualine/lualine.nvim](https://github.com/nvim-lualine/lualine.nvim)
- [nvim-telescope/telescope.nvim](https://github.com/nvim-telescope/telescope.nvim)
- [nvim-treesitter/nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter)
- [phaazon/hop.nvim](https://github.com/phaazon/hop.nvim)
- [projekt0n/circles.nvim](https://github.com/projekt0n/circles.nvim)
- [rcarriga/nvim-notify](https://github.com/rcarriga/nvim-notify)
- [romgrk/barbar.nvim](https://github.com/romgrk/barbar.nvim)

## Inspirations

- [Binaryify/OneDark-Pro](https://github.com/Binaryify/OneDark-Pro)
- [monsonjeremy/onedark.nvim](https://github.com/monsonjeremy/onedark.nvim)
- [folke/tokyonight.nvim](https://github.com/folke/tokyonight.nvim)
- [eddyekofo94/gruvbox-flat.nvim](https://github.com/eddyekofo94/gruvbox-flat.nvim)

## Screenshot

- Font: [JetBrains Mono](https://www.jetbrains.com/lp/mono/)
- Terminal: [kitty](https://sw.kovidgoyal.net/kitty)
- dotfiles: [ful1e5/dotfiles](https://github.com/ful1e5/dotfiles/tree/main/nvim/.config/nvim)

<p align="center">
  <img src="https://imgur.com/1ksV9DW.png" alt="onedark" />
</p>

### Default

<p align="center">
  <img src="https://imgur.com/0R7dvzB.png" alt="Default" />
</p>

### Normal

```vim
" VimScript
let g:onedark_comment_style = "NONE"
let g:onedark_keyword_style = "NONE"
let g:onedark_function_style = "NONE"
let g:onedark_variable_style = "NONE"
```

```lua
-- Lua
require("onedark").setup({
  comment_style = "NONE",
  keyword_style = "NONE",
  function_style = "NONE",
  variable_style = "NONE"
  -- ... your onedark config
})
```

<p align="center">
  <img src="https://imgur.com/ujZ5tkS.png" alt="Normal" />
</p>

### Italic

```vim
" VimScript
let g:onedark_comment_style = "italic"
let g:onedark_keyword_style = "italic"
let g:onedark_function_style = "italic"
let g:onedark_variable_style = "italic"
```

```lua
-- Lua
require("onedark").setup({
  comment_style = "italic",
  keyword_style = "italic",
  function_style = "italic",
  variable_style = "italic"
  -- ... your onedark config
})
```

<p align="center">
  <img src="https://imgur.com/YQYPpn1.png" alt="Italic" />
</p>

### Minimal

```vim
" VimScript
let g:onedark_hide_inactive_statusline = 1
let g:onedark_dark_sidebar = 0
let g:onedark_dark_float = 0
```

```lua
-- Lua
require("onedark").setup({
  hide_inactive_statusline = true,
  dark_sidebar = false,
  dark_float = false
  -- ... your onedark config
})
```

<p align="center">
  <img src="https://imgur.com/UuEbkyZ.png" alt="Minimal" />
</p>

### Highlight Line-Number

```vim
" VimScript
let g:onedark_highlight_linenumber = 1
```

```lua
-- Lua
require("onedark").setup({
  highlight_linenumber = true
  -- ... your onedark config
})
```

<p align="center">
  <img src="https://imgur.com/ZAiFAAM.png" alt="Highlight Line-Number" />
</p>

### Telescope

<p align="center">
  <img src="https://imgur.com/xLKjj0I.png" alt="Telescope plugin" />
</p>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [unreleased]

### Added

- plugin support [ggandor/lightspeed.nvim](https://github.com/ggandor/lightspeed.nvim) #103
- feat: use native autocmds on nvim >=0.7
- feat: use typing (types.od.HighlightStyle) inside `theme.lua`
- feat: add non-exists highlight from `overrides` option in dev mode (fixed #99)
- enhance: bright & dark colors updated in terminal themes
- chore: unlink 'Visual' highlight

### Changed

- fixed #101
- fixed #104
- refactor: implement override func in `util.load`
- docs: fixed config's default value in README.md
- fixed projekt0n/github-nvim-theme#172
- fixed #106
- fix: use pcall to clear autocmd silently projekt0n/github-nvim-theme#178 ( @xiyaowong )
- fixed #108
- fixed: pcall bad argument inside autocmds module

## [v0.0.4]- 21 Jan 2022

### Added

- plugin support [hop.nvim](https://github.com/phaazon/hop.nvim)
- lualine `terminal` highlight added
- plugin support [coc.nvim](https://github.com/neoclide/coc.nvim)
- define global in project `luarc` config file.
- config option added `hide_end_of_buffer`
- allow customizing highlight group #91
- Removed unnecessary defer function & highlights
- highlight added for `GitSignsCurrentLineBlame`
- feat: `githooks` init
- feat: `types` module init
- cleanup: removed 'transform_colors' option from terminal themes generation
- feat: 'ConfigSchema' type init
- feat: `od.Highlights.*` type added
- refactor: `config` module
- refactor: `config.config` table renamed to `config.schema`
- refactor: `util` module
- breaking change: color palette redefine
- feat: `require("onedark").get_colors` added
- fix: autocompletion color in nvim in-built `terminal` #95

### Changed

- `hide_inactive_statusline` option deprecated from `lualine` plugin
- `javascriptTSTagAttribute` highlight removed
- `TSTagAttribute` highlight added
- yellow changed to the same version as onedark palette. monsonjeremy/onedark.nvim#19
- Boolean set to orange as in VSC onedark theme. monsonjeremy/onedark.nvim#19
- refactor override function to util.lua module #91
- formatter changed `LuaFormatter` -> `Stylua`
- refactor: modules export
- refactor: 'util' and 'extra' themes
- refactor: use `TerminalConf` type in terminal theme generation
- refactor: 'nvim-dev-icons' palette
- refactor: module renamed `extra` -> `terminal`
- docs: re-arrange heading section & plugins list
- fix: alacritty `black` color in bright section
- nvim-dev-icons colors fixed #98 projekt0n/circles.nvim#6

## [v0.0.3]- 11 Dec 2021

### Added

- docs updated
- Assign `dark5` color to `LineNr` inside `transparent` mode
- docs: contributing guidelines added
- docs: code of conduct added
- feat: `highlight_linenumber` config for setting dark color to number line
- feat: support `highlight_linenumber` with gitsigns
- enhance: nvim-cmp highlight
- change popup 'select' highlight background to green
- enhance(nvim-cmp):separate match highlights
- plugin-support: [nvim-notify](https://github.com/rcarriga/nvim-notify)
- enhance: set default config value to `vim.g` variables
- feat: `onedark_transparent_sidebar` option #60
- docs: vim config docs added inside README.md
- feat: store `user_config` value to `vim.g` variables
- docs: luadoc added inside `config` and `extra` module
- A new lualine theme added 'onedark-nvim' #65
- expose onedark theme in [fzf.vim](https://github.com/junegunn/fzf.vim)
- added hot reloaded to `onedark-nvim` lualine theme
- feat: `lualine_bold` option added
- luadoc: fix warning inside config module
- fixed #71 in normal mode
- vscode like `nvim-cmp` highlights [ref:nvim-cmp wiki](https://github.com/hrsh7th/nvim-cmp/wiki/Menu-Appearance#how-to-add-visual-studio-code-dark-theme-colors-to-the-menu)
- `nvim-cmp` all lspkind supported
- Link common `diff` highlights group
- dark foreground color for visited entries in telescope.nvim

### Changed

- projekt0n/github-nvim-theme#73 fixed
- refactor: lualine colors added inside `colors.lua`
- breaking change: consistent variable naming in `onedark` config
- undefined color `syntax.comment` fixed
- lualine theme provided by official lualine repo (fix #46)
- refactor: group colors according to assignment
- enhance: barbar highlighting monsonjeremy/onedark.nvim#16
- enhance: 'cmp' menu with 'fg' color
- docs: minimal screenshot updated in [README.md#minimal](https://github.com/ful1e5/onedark.nvim#minimal)
- docs: removed emoji from README.md
- cleanup: removed vim config deprecated message
- breaking change: use snake_case inside config and alacritty module
- breaking change: changed config variables name inside `vim.g` variables
- refactor: colors transform config inside `extra` module
- fix: diff text with comment text #68
- docs: cleanup some nonsense
- `NvimTreeGitRenamed` highlight with onedark purple color

## [v0.0.2]- 19 Sep 2021

### Added

- **Warning** on `vim.g` configuration
- Configure plugin in lua using `require('onedark').setup({})` **[fix monsonjeremy/onedark.nvim#5]**
- Minimal config example added inside [README.md#minimal](./README.md#minimal)
- Linting Github Action init
- Format .lua code with [lua-format](https://github.com/Koihik/LuaFormatter)
- **kitty tab** colors are changed
- Init: `Makefile, .lua-format, .luacheckrc`
- Add "NONE" color compatibility for colors override
- `barbar.nvim` plugin support. related to monsonjeremy/onedark.nvim#8
- `glyph-palette.vim` plugin support
- kitty border color added
- `./colors/onedark.vim` code in lua
- [nvim-web-devicons](https://github.com/kyazdani42/nvim-web-devicons) plugin support
- `msg_area_style` config added
- `kitty` target added inside `Makefile` for reloading kitty theme
- `TSVariable` & `TSOperator` colors added inside `colors.lua`

lua/onedark/util.lua
local hsluv = require('onedark.hsluv')
local types = require('onedark.types')

---@class od.Util
local util = {}

---@type table<number, od.HexColor>
util.used_colors = {}

---@type table<number, od.HexColor>
util.color_cache = {}

---@type od.HexColor
util.bg = '#000000'

---@type od.HexColor
util.fg = '#ffffff'

---Convert HexColor to RGB
---@param hex_str od.HexColor
---@return table<number, number>
local hex_to_rgb = function(hex_str)
  local hex = '[abcdef0-9][abcdef0-9]'
  local pat = string.format('^#(%s)(%s)(%s)$', hex, hex, hex)
  hex_str = string.lower(hex_str)

  assert(string.find(hex_str, pat) ~= nil, 'hex_to_rgb: invalid hex_str: ' .. tostring(hex_str))

  local r, g, b = string.match(hex_str, pat)
  return { tonumber(r, 16), tonumber(g, 16), tonumber(b, 16) }
end

---@param fg od.HexColor foreground color
---@param bg od.HexColor background color
---@param alpha number number between 0 and 1. 0 results in bg, 1 results in fg
---@return od.HexColor
util.blend = function(fg, bg, alpha)
  bg = hex_to_rgb(bg)
  fg = hex_to_rgb(fg)

  local blend_channel = function(i)
    local ret = (alpha * fg[i] + ((1 - alpha) * bg[i]))
    return math.floor(math.min(math.max(0, ret), 255) + 0.5)
  end

  return string.format('#%02X%02X%02X', blend_channel(1), blend_channel(2), blend_channel(3))
end

---@param hex od.HexColor Color
---@param amount number number between 0 and 1. 0 results in bg, 1 results in fg
---@param bg od.HexColor Background color
---@return od.HexColor
util.darken = function(hex, amount, bg)
  return util.blend(hex, bg or util.bg, math.abs(amount))
end

---@param hex od.HexColor Color
---@param amount number number between 0 and 1. 0 results in bg, 1 results in fg
---@param fg od.HexColor Foreground color
---@return od.HexColor
util.lighten = function(hex, amount, fg)
  return util.blend(hex, fg or util.fg, math.abs(amount))
end

---@param color od.HexColor
---@param percentage number
---@return od.HexColor
util.brighten = function(color, percentage)
  local hsl = hsluv.hex_to_hsluv(color)
  local larpSpace = 100 - hsl[3]
  if percentage < 0 then
    larpSpace = hsl[3]
  end
  hsl[3] = hsl[3] + larpSpace * percentage
  return hsluv.hsluv_to_hex(hsl)
end

---@param color od.HexColor
---@return od.HexColor
util.invert_color = function(color)
  if color ~= 'NONE' then
    local hsl = hsluv.hex_to_hsluv(color)
    hsl[3] = 100 - hsl[3]
    if hsl[3] < 40 then
      hsl[3] = hsl[3] + (100 - hsl[3]) * 0.3
    end
    return hsluv.hsluv_to_hex(hsl)
  end
  return color
end

---Dump unused colors
---@param colors od.ColorPalette
util.debug = function(colors)
  for name, color in pairs(colors) do
    if type(color) == 'table' then
      util.debug(color)
    else
      if util.used_colors[color] == nil then
        print(string.format('not used color: %s=%s', name, color))
      end
    end
  end
end

---@param color od.HexColor
---@return od.HexColor
util.get_color = function(color)
  if vim.o.background == 'dark' then
    return color
  end
  if not util.color_cache[color] then
    util.color_cache[color] = util.invert_color(color)
  end
  return util.color_cache[color]
end

---@param hi_name string
---@param hi od.Highlight|od.LinkHighlight
util.highlight = function(hi_name, hi)
  if hi.fg then
    util.used_colors[hi.fg] = true
  end
  if hi.bg then
    util.used_colors[hi.bg] = true
  end
  if hi.sp then
    util.used_colors[hi.sp] = true
  end

  local style = hi.style and 'gui=' .. hi.style or 'gui=NONE'
  local fg = hi.fg and 'guifg=' .. hi.fg or 'guifg=NONE'
  local bg = hi.bg and 'guibg=' .. hi.bg or 'guibg=NONE'
  local sp = hi.sp and 'guisp=' .. hi.sp or ''

  local hi_cmd = string.format('highlight %s %s %s %s %s', hi_name, style, fg, bg, sp)

  if hi.link then
    vim.cmd(string.format('highlight! link %s %s', hi_name, hi.link))
  else
    -- local data = {}
    -- if color.fg then data.foreground = color.fg end
    -- if color.bg then data.background = color.bg end
    -- if color.sp then data.special = color.sp end
    -- if color.style then data[color.style] = true end
    -- vim.api.nvim_set_hl(ns, group, data)
    vim.cmd(hi_cmd)
  end
end

---@param base od.Highlights.Base
util.syntax = function(base)
  for hi_name, hi in pairs(base) do
    util.highlight(hi_name, hi)
  end
end

---@param colors od.ColorPalette
util.terminal = function(colors)
  -- dark
  vim.g.terminal_color_0 = colors.black0
  vim.g.terminal_color_8 = colors.fg_dark
  -- light
  vim.g.terminal_color_7 = colors.fg_dark
  vim.g.terminal_color_15 = colors.fg0
  -- colors
  vim.g.terminal_color_1 = colors.red0
  vim.g.terminal_color_9 = colors.red1
  vim.g.terminal_color_2 = colors.green0
  vim.g.terminal_color_10 = colors.green0
  vim.g.terminal_color_3 = colors.yellow0
  vim.g.terminal_color_11 = colors.yellow1
  vim.g.terminal_color_4 = colors.blue0
  vim.g.terminal_color_12 = colors.blue1
  vim.g.terminal_color_5 = colors.purple0
  vim.g.terminal_color_13 = colors.purple0
  vim.g.terminal_color_6 = colors.cyan0
  vim.g.terminal_color_14 = colors.cyan0

  if vim.o.background == 'light' then
    for i = 0, 15, 1 do
      vim.g['terminal_color_' .. i] = util.get_color(vim.g['terminal_color_' .. i])
    end
  end
end

---@param colors od.ColorPalette
---@return od.ColorPalette
util.light_colors = function(colors)
  if type(colors) == 'string' then
    return util.get_color(colors)
  end
  local ret = {}
  for key, value in pairs(colors) do
    ret[key] = util.light_colors(value)
  end
  return ret
end

---Override custom highlights in `group`
---@param group table
---@param overrides table
---@param force boolean
util.apply_overrides = function(group, overrides, force)
  for k, v in pairs(overrides) do
    if type(v) == 'table' and group[k] ~= nil or force then
      group[k] = v
    end
  end
end

---@param hi od.Highlights
util.load = function(hi)
  vim.cmd('hi clear')
  if vim.fn.exists('syntax_on') then
    vim.cmd('syntax reset')
  end

  vim.o.termguicolors = true
  vim.g.colors_name = 'onedark'

  -- override colors
  local overrides = hi.config.overrides(hi.colors)
  util.apply_overrides(hi.base, overrides, hi.config.dev)
  util.apply_overrides(hi.plugins, overrides, hi.config.dev)

  local autocmds = require('onedark.autocmds')

  -- load base theme
  util.syntax(hi.base)
  autocmds.set(hi.config)
  util.terminal(hi.colors)
  util.syntax(hi.plugins)
end

---@param colors od.ColorPalette
---@param oride_colors od.ColorPalette
---@return od.ColorPalette
function util.color_overrides(colors, oride_colors)
  if type(oride_colors) == 'table' then
    for key, value in pairs(oride_colors) do
      local err_msg = string.format('color "%s" does not exist', key)

      if not colors[key] then
        error(err_msg)
      end
      -- Patch: https://github.com/ful1e5/onedark.nvim/issues/6
      if type(colors[key]) == 'table' then
        util.color_overrides(colors[key], value)
      else
        if value:lower() == types.od.HighlightStyle.None:lower() then
          -- set to none
          colors[key] = types.od.HighlightStyle.None
        elseif string.sub(value, 1, 1) == '#' then
          -- hex override
          colors[key] = value
        else
          -- another group
          if not colors[value] then
            error(err_msg)
          end
          colors[key] = colors[value]
        end
      end
    end
  end
  return colors
end

-- Simple string interpolation.
--
-- Example template: "${name} is ${value}"
--
---@param str string template string
---@param table table key value pairs to replace in the string
util.template = function(str, table)
  return (str:gsub('($%b{})', function(w)
    return table[w:sub(3, -2)] or w
  end))
end

return util
